sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -543,6 +543,12 @@ object ScalaReflection extends ScalaReflection {
             throw new UnsupportedOperationException(s"`$fieldName` is a reserved keyword and " +
               "cannot be used as field name\n" + walkedTypePath)
           }
+          // It is ok to use head here because a variable name needs to be non empty
+          if (fieldName.head.isDigit) {
```

Comment:
I would leverage something like `Character.isJavaIdentifierStart` though.

Reply:
I like that a lot, it catches other issues besides numbers like strange punctuations and so on. Thanks for the suggestion @HyukjinKwon